src/sentry/utils/services.py

```diff
@@ -76,3 +83,173 @@ def expose(self, context):
+    to any registered backend, the function will raise a ``InvalidBackend``
+    exception. If any referenced secondary backends are not registered names,
+    they will be discarded and logged.
+
+    The selector function and callback function are both provided as paths to a
+    callable that will be imported at backend instantation.
+
+    Implementation notes:
+
+    - Only method access is delegated to the individual backends. Attribute
+      values are returned from the base backend. Only methods that are defined
+      on the base backend are eligble for delegation (since these methods are
+      considered the public API.)
+    - The backend makes no attempt to synchronize common backend option values
+      between backends (e.g. TSDB rollup configuration) to ensure equivalency
+      of request parameters based on configuration.
+    - Each backend is associated with an executor pool which defaults to a
+      thread pool implementation unless otherwise specified in the backend
+      configuration. If the backend itself is not thread safe (due to socket
+      access, etc.), it's recommended to specify a pool size of 1 to ensure
+      exclusive access to resources. Each executor is started when the first
+      task is submitted.
+    - The request is added to the request queue of the primary backend using a
+      blocking put. The request is added to the request queue(s) of the
+      secondary backend(s) as a non-blocking put (if these queues are full, the
+      request is rejected and the future will raise ``Queue.Full`` when
+      attempting to retrieve the result.)
+    - The ``callback_func`` is called after all futures have completed, either
+      successfully or unsuccessfully. The function is parameters are the method
+      name, calling arguments (as returned by ``inspect.getcallargs``) and a
+      mapping of backend name to ``Future`` objects.
+    """
+
+    class InvalidBackend(Exception):
+        """\
+        Exception raised when an invalid backend is returned by a selector
+        function.
+        """
+
+    def __init__(self, backend_base, backends, selector_func, callback_func=None):
+        if callback_func is not None:
+            callback_func = import_string(callback_func)
+
+        self.__backend_base = import_string(backend_base)
+
+        def load_executor(options):
+            path = options.get('path')
+            if path is None:
+                executor_cls = ThreadedExecutor
+            else:
+                executor_cls = import_string(path)
+            return executor_cls(**options.get('options', {}))
+
+        self.__backends = {}
+        for name, options in backends.items():
+            self.__backends[name] = (
+                import_string(options['path'])(**options.get('options', {})),
+                load_executor(options.get('executor', {})),
+            )
+
+        self.__selector_func = import_string(selector_func)
+        self.__callback_func = callback_func
```

Comment:
It might make sense for me to just rewrite this as `self.__callback_func = import_string(callback_func) if callback_func is not None else None` for symmetry's sake and drop L168-170?

Reply:
I will admit I was looking for where `callback_func` was going to be used above, so couldn't hurt to bring them closer at minimum.

Reply:
Updated here: c5cc4efff0bc9459ffb9c33acf5012943aa6794e.